### PR TITLE
Fixes [#276] Throw a nicer error when the routing services haven't been added to DI

### PR DESCRIPTION
**Pull request by javiercn**

### CODE REVIEW

File: test/Microsoft.AspNet.Routing.Tests/BuilderExtensionsTest.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.Builder;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNet.Routing.Tests
```
Comment (dougbu) [inline test/Microsoft.AspNet.Routing.Tests/BuilderExtensionsTest.cs]: `Microsoft.AspNet.Builder` to match `BuilderExtensions` (plus: almost never `...Test` or `...Tests`)

File: test/Microsoft.AspNet.Routing.Tests/BuilderExtensionsTest.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            // Arrange
+            var applicationBuilderMock = new Mock<IApplicationBuilder>();
+            applicationBuilderMock
+                .Setup(s => s.ApplicationServices)
+                .Returns(Mock.Of<IServiceProvider>());
+
+            var router = Mock.Of<IRouter>();
+            
+            // Act
```
Comment (dougbu) [inline test/Microsoft.AspNet.Routing.Tests/BuilderExtensionsTest.cs]: `// Act & Assert` and get rid of extra `// Assert` line below. Same for other tests.

File: test/Microsoft.AspNet.Routing.Tests/BuilderExtensionsTest.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            var error = Assert.Throws<InvalidOperationException>(
```
Comment (dougbu) [inline test/Microsoft.AspNet.Routing.Tests/BuilderExtensionsTest.cs]: We usually use `ex` or `exception` for similar variables.

**Comment by dougbu** [comment]
:watch:

**Comment by javiercn** [comment]
@dougbu updated

File: src/Microsoft.AspNet.Routing/RouteBuilder.cs [hunk 4]
```diff
@@ -21,6 +23,14 @@ public RouteBuilder(IApplicationBuilder applicationBuilder, IRouter defaultHandl
                 throw new ArgumentNullException(nameof(applicationBuilder));
             }
 
+            if (applicationBuilder.ApplicationServices.GetService(typeof(RoutingMarkerService)) == null)
+            {
+                throw new InvalidOperationException(Resources.FormatUnableToFindServices(
```
Comment (dougbu) [inline src/Microsoft.AspNet.Routing/RouteBuilder.cs]: Forgot to ask last time: Do we really want the constructor to `throw` for a situation unrelated to its parameters? Realize `RouteBuilder` objects aren't passed into the system but am still hoping there's another place where we can cover the non-`UseRouting()` scenario.

File: test/Microsoft.AspNet.Routing.Tests/RouteBuilderTest.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        [Fact]
+        public void Ctor_ThrowsInvalidOperationException_IfRoutingMarkerServiceIsNotRegistered()
+        {
+            // Arrange
+            var applicationBuilderMock = new Mock<IApplicationBuilder>();
+            applicationBuilderMock
+                .Setup(s => s.ApplicationServices)
+                .Returns(Mock.Of<IServiceProvider>());
+
+            // Act
```
Comment (dougbu) [inline test/Microsoft.AspNet.Routing.Tests/RouteBuilderTest.cs]: Missed an `// Act & Assert`

File: test/Microsoft.AspNet.Routing.Tests/RouteBuilderTest.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+            var error = Assert.Throws<InvalidOperationException>(() => new RouteBuilder(applicationBuilderMock.Object));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Routing.Tests/RouteBuilderTest.cs]: `error` -> `exception` as in other new test

**Comment by dougbu** [comment]
:watch: for small touch-ups and one [open question](https://github.com/aspnet/Routing/pull/278#discussion-diff-49766388).

**Comment by javiercn** [comment]
@dougbu updated. @rynowak any comment?

File: src/Microsoft.AspNet.Routing/RouteBuilder.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Routing/RouteBuilder.cs]: Caught @Eilon and he's :ok: with this. @rynowak has more to say however.

**Comment by dougbu** [comment]
:shipit: but definitely get input from @rynowak.

File: src/Microsoft.AspNet.Routing/RouteBuilder.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/RouteBuilder.cs]: Hmm - what's the reason for this?  Like, what's the callstack that needs to throw that isn't throwing?

**Comment by javiercn** [comment]
@rynowak Updated to include your suggestions

File: src/Microsoft.AspNet.Routing/BuilderExtensions.cs [hunk 7]
```diff
@@ -29,7 +31,51 @@ public static IApplicationBuilder UseRouter(this IApplicationBuilder builder, IR
...
+
+            if (builder.ApplicationServices.GetService(typeof(RoutingMarkerService)) == null)
+            {
+                throw new InvalidOperationException(Resources.FormatUnableToFindServices(
+                    nameof(IServiceCollection),
+                    nameof(RoutingServiceCollectionExtensions.AddRouting),
+                    "ConfigureServices(...)"));
+            }
+
+            return new RouteBuilder(builder, defaultHandler);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Routing/BuilderExtensions.cs]: We should also add the middleware to the pipeline here. This is a `Use` method after all.

Comment (rynowak) [inline src/Microsoft.AspNet.Routing/BuilderExtensions.cs]: Yeah let's leave this.

**Comment by rynowak** [comment]
:shipit: for https://github.com/aspnet/Routing/commit/db34f85ea1ef69cf980a633db260ce7d889bf662

**Pull request merged**